packages/dds/tree/src/feature-libraries/default-schema/schemaChecker.ts

```diff
@@ -32,48 +32,58 @@ export enum SchemaValidationError {
 }
 
 /**
- * Throws a UsageError if maybeError indicates a tree is out of schema.
+ * Throws a UsageError indicating a tree is out of schema.
  */
-export function inSchemaOrThrow(maybeError: SchemaValidationError | undefined): void {
-	if (maybeError !== undefined) {
-		throw new UsageError(
-			`Tree does not conform to schema: ${SchemaValidationError[maybeError]}`,
-		);
-	}
+export function throwOutOfSchema(maybeError: SchemaValidationError): never {
+	throw new UsageError(
+		`Tree does not conform to schema: ${SchemaValidationError[maybeError]}`,
+	);
 }
 
+type NotUndefined = number | string | boolean | bigint | symbol | object;
+
 /**
  * Deeply checks that the provided node complies with the schema based on its identifier.
+ *
+ * @param onError - Called with the first error (if any).
+ *
+ * @returns the return value from `onError` if the node or anything inside of it is out of schema, otherwise `undefined`.
  */
-export function isNodeInSchema(
+export function isNodeInSchema<T extends NotUndefined>(
 	node: MinimalMapTreeNodeView,
 	schemaAndPolicy: SchemaAndPolicy,
-): SchemaValidationError | undefined {
+	onError: (error: SchemaValidationError) => T,
```

Comment:
Changes to this file were just to improve the debugging experience. Now it's possible for the out of schema error to be thrown with the exact location of the tree that's out of schema observable in the call stack.